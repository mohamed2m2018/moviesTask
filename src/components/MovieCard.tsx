import React from 'react'
import styled from 'styled-components/native'
import { perfectHeight, perfectWidth } from '../helpers/commonFunctions'
import Genres from './Genres'

interface Props {
  title: string,
  date: string,
  genres: string[],
  uri: string,
  rating: string,
  onPress: () => void
}

const MovieCard = ({ title, date, uri, rating, genres, onPress }: Props) => {
  return (
    <OuterContainer onPress={onPress}>
      <FirstColumn>
        <Image resizeMode={'cover'} source={{ uri: 'https://image.tmdb.org/t/p/original/' + uri }} />
      </FirstColumn>
      <MiddleColumn>
        <MainText>{title}</MainText>
        <DateText>{date}</DateText>
        <Row>
          {
            genres.map((name: string, index: number) => <Genres key={index} text={name} />)
          }
        </Row>
      </MiddleColumn>
      <LastColumn>
        <RatingText>{rating}%</RatingText>
      </LastColumn>
    </OuterContainer>
  )
}

export default MovieCard

const OuterContainer = styled.TouchableOpacity`
  flex-direction: row;
  justify-content: space-between;
  align-self: center;
  margin-vertical: ${perfectHeight(10)}px;
  padding: ${perfectHeight(10)}px;
  border-radius: 10px;
  background-color: white;
  width: ${perfectWidth(350)}px;
`
const Row = styled.View`
  flex-direction: row;
  flex-wrap: wrap;
`
const FirstColumn = styled.View`
  flex-direction: column;
  width: ${perfectWidth(120)}px;
`


const MiddleColumn = styled.View`
  flex-direction: column;
  width: ${perfectWidth(160)}px;
`

const LastColumn = styled.View`
  flex-direction: column;
  justify-content: flex-end;
`
const Image = styled.Image`
  height: ${perfectHeight(120)}px;
  width: ${perfectWidth(100)}px;
  border-radius: 10px;
`
const MainText = styled.Text`
  font-size: 20px;
  margin-bottom: 10px;
  font-weight: bold;
`
const DateText = styled.Text`
  font-size: 15px;
  margin-bottom: 20px;
`

const RatingText = styled.Text`
  color: #73F340;
  font-size: 20px;
  font-weight: bold;
`