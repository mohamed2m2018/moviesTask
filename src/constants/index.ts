export const colors = {
  grey: '#d8d8d8',
  green: '#59C14C',
  white: '#fff',
  black: '#000',
  background: '#F6F6F8',
};

export const testIds = {
  MOVIE_CARD_TITLE: 'MOVIE_CARD_TITLE',
  MOVIE_CARD_DATE: 'MOVIE_CARD_DATE',
  MOVIE_CARD_RATING: 'MOVIE_CARD_RATING',
  MOVIE_CARD_TOUCHABLE: 'MOVIE_CARD_TOUCHABLE',
  MOVIES_FLATLIST: 'MOVIES_FLATLIST',
};
