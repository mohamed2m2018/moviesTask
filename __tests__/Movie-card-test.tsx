import 'react-native';
import React from 'react';

// Note: test renderer must be required after react-native.
import renderer, {act} from 'react-test-renderer';
import MovieCard from '../src/components/MovieCard';
import {fireEvent, render} from '@testing-library/react-native';
import {testIds} from '../src/constants';

let props = {
  title: 'No time to die',
  date: '29-10-2021',
  genres: ['Thriller', 'Action'],
  uri: '/iOPXG7q6ZqxSFp2iSY2CvgeRXyC.jpg',
  rating: '79',
  onPress: () => {},
};

describe('Movie card', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  it('renders correctly', () => {
    const screenToBeSnapped = renderer
      .create(<MovieCard {...props} />)
      .toJSON();
    expect(screenToBeSnapped).toMatchSnapshot();
  });

  it('renders correct title', () => {
    const {getByTestId} = render(<MovieCard {...props} />);
    const title = getByTestId(testIds.MOVIE_CARD_TITLE);
    expect(title.children[0]).toEqual(props.title);
  });
  it('renders correct rating', () => {
    const {getByTestId} = render(<MovieCard {...props} />);
    const rating = getByTestId(testIds.MOVIE_CARD_RATING);
    expect(rating.children[0]).toEqual(props.rating);
  });
  it('renders correct date', () => {
    const {getByTestId} = render(<MovieCard {...props} />);
    const date = getByTestId(testIds.MOVIE_CARD_DATE);
    expect(date.children[0]).toEqual(props.date);
  });
  it('fires onPress prop when clicked on', async () => {
    const MethodToBeSent = jest.fn(() => {});
    props.onPress = MethodToBeSent;
    const {getByTestId} = render(<MovieCard {...props} />);
    const touchable = getByTestId(testIds.MOVIE_CARD_TOUCHABLE);
    await act(async () => {
      fireEvent.press(touchable);
    });
    expect(MethodToBeSent).toHaveBeenCalled();
  });
});
