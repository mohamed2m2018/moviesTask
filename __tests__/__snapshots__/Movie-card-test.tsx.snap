// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie card renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignSelf": "center",
      "backgroundColor": "white",
      "borderRadius": 10,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginVertical": 20.523076923076925,
      "opacity": 1,
      "paddingBottom": 20.523076923076925,
      "paddingLeft": 20.523076923076925,
      "paddingRight": 20.523076923076925,
      "paddingTop": 20.523076923076925,
      "width": 700,
    }
  }
  testID="MOVIE_CARD_TOUCHABLE"
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "column",
          "width": 240,
        },
      ]
    }
  >
    <Image
      resizeMode="cover"
      source={
        Object {
          "uri": "https://image.tmdb.org/t/p/original//iOPXG7q6ZqxSFp2iSY2CvgeRXyC.jpg",
        }
      }
      style={
        Array [
          Object {
            "borderRadius": 10,
            "height": 246.27692307692305,
            "width": 200,
          },
        ]
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "column",
          "width": 320,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontSize": 20,
            "fontWeight": "bold",
            "marginBottom": 10,
          },
        ]
      }
      testID="MOVIE_CARD_TITLE"
    >
      No time to die
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontSize": 15,
            "marginBottom": 20,
          },
        ]
      }
      testID="MOVIE_CARD_DATE"
    >
      29-10-2021
    </Text>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#d8d8d8",
              "borderRadius": 10,
              "marginBottom": 10.261538461538462,
              "marginRight": 10,
              "paddingHorizontal": 10,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000",
              },
            ]
          }
        >
          Thriller
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#d8d8d8",
              "borderRadius": 10,
              "marginBottom": 10.261538461538462,
              "marginRight": 10,
              "paddingHorizontal": 10,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000",
              },
            ]
          }
        >
          Action
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "column",
          "justifyContent": "flex-end",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#59C14C",
            "fontSize": 20,
            "fontWeight": "bold",
          },
        ]
      }
      testID="MOVIE_CARD_RATING"
    >
      79
      %
    </Text>
  </View>
</View>
`;
